File: javascript/creational/factory.js
class Vehicle {
  constructor(name) {
    this.name = name;
  }

  getName = () => {
      return this.name;
  }
}

class Car extends Vehicle {
  constructor(name) {
    super(name);
  }

  wheels = () => {
    return 2;
  };
}

class Truck extends Vehicle {
  constructor(name) {
    super(name);
  }

  wheels = () => {
    return 4;
  };
}

class AbstractFactory {
  #type = "Car";
  #origin;
  #date;
  #vehicle;

  constructor(type) {
    this.#type = type;
  }

  produce = (origin) => {
    this.#origin = origin;
    this.#date = new Date().getHours();
    let description = `${this.#type} from ${this.#origin} produced at: ${
      this.#date
    }`;
    if (this.#type === "Car") {
      this.#vehicle = new Car(description);
    } else if (this.#type === "Truck") {
      this.#vehicle = new Truck(description);
    } else {
        throw new Error("Abstract Factory does not know how to produce a vehicle");
    }
  };

  getType() {
    return this.#type;
  }

  getDate() {
    return this.#date;
  }

  getOrigin() {
    return this.#origin;
  }

  getVehicle() {
    return this.#vehicle;
  }
}

class HcmCarFactory extends AbstractFactory {
  constructor() {
    super("Car");
  }

  produce(origin) {
    return super.produce(origin);
  }
}

class HnTruckFactory extends AbstractFactory {
  constructor() {
    super("Truck");
  }

  produce(origin) {
    return super.produce(origin);
  }
}

module.exports = {
    Truck,
    Car,
    HcmCarFactory,
    HnTruckFactory
}

File: go/creational/factory.go
package creational

import (
	"errors"
	"strconv"
)

type Vehicle interface {
	GetName() string
}

type Car struct {
	Name       string
	WheelCount int
}

type Truck struct {
	Name string
}

type Factory struct {
	Name string
}

func (c *Car) GetName() string {
	return c.Name + strconv.Itoa(c.WheelCount)
}

func (t *Truck) GetName() string {
	return t.Name
}

func (f *Factory) GetVehicle(v string) (Vehicle, error) {
	switch v {
	case "car":
		return &Car{Name: "Car", WheelCount: 4}, nil
	case "truck":
		return &Truck{Name: "Truck"}, nil
	default:
		return nil, errors.New("Vehicle not found")
	}
}
